ord_schema/resolvers.py

```diff
@@ -0,0 +1,188 @@
+        for identifier in compound.identifiers:
+            if identifier.type == identifier.NAME:
+                try:
+                    smiles, resolver = name_resolve('name', identifier.value)
+                    new_identifier = compound.identifiers.add()
+                    new_identifier.type = new_identifier.SMILES
+                    new_identifier.value = smiles
+                    new_identifier.details = f'NAME resolved by the {resolver}'
+                    modified = True
+                    break
+                except ValueError:
+                    pass
+    return modified
+
+
+def _pubchem_resolve(value_type, value):
+    """Resolves compound identifiers to SMILES via the PubChem REST API."""
+    response = urllib.request.urlopen(
+        f'https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/{value_type}/'
+        f'{urllib.parse.quote(value)}/property/IsomericSMILES/txt')
+    return response.read().decode().strip()
+
+
+def _cactus_resolve(value_type, value):
+    """Resolves compound identifiers to SMILES via the CACTUS API."""
+    del value_type
+    response = urllib.request.urlopen(
+        'https://cactus.nci.nih.gov/chemical/structure/'
+        f'{urllib.parse.quote(value)}/smiles')
+    return response.read().decode().strip()
+
+
+def _emolecules_resolve(value_type, value):
+    """Resolves compound identifiers to SMILES via the eMolecules API."""
+    del value_type
+    response = urllib.request.urlopen(
+        'https://www.emolecules.com/lookup?q={}'.format(
+            urllib.parse.quote(value)))
+    response_text = response.read().decode().strip()
+    if response_text == '__END__':
+        raise urllib.error.HTTPError(None, None,
+                                     'eMolecules lookup unsuccessful', None,
+                                     None)
+    return response_text.split('\t')[0]
+
+
+def resolve_input(input_string):
+    """Resolve a text-based description of an input in one of the following
+    formats:
+        (1) [AMOUNT] of [NAME]
+        (2) [AMOUNT] of [CONCENTRATION] [SOLUTE] in [SOLVENT]
+
+    Args:
+        input_string: String describing the input.
+
+    Returns:
+        ReactionInput message.
+
+    Raises:
+        ValueError: if the string cannot be parsed properly.
+    """
+    reaction_input = reaction_pb2.ReactionInput()
+    if ' of ' not in input_string:
+        raise ValueError('String does not match template!')
+    amount_string, description = input_string.split(' of ')
+    if ' in ' not in description:
+        component_name = description
+        component = reaction_input.components.add()
+        component.CopyFrom(
+            message_helpers.build_compound(name=component_name.strip(),
+                                           amount=amount_string))
+        resolve_names(reaction_input)
+        return reaction_input
+    pattern = re.compile(r'(\d+.?\d*)\s*(\w+) (.*) in (.*)')
```

Comment:
This looks ok but I'd probably use `(.+)` instead of `(.*)`. I might also replace `\s*` with `\s?` so you only allow a maximum of one space before the units.

Reply:
Done